src/core/main.c

```diff
@@ -1253,37 +1253,39 @@ static void bump_file_max_and_nr_open(void) {
 #endif
 
 #if BUMP_PROC_SYS_FS_NR_OPEN
-        int v = INT_MAX;
-
-        /* Argh! The kernel enforces maximum and minimum values on the fs.nr_open, but we don't really know
-         * what they are. The expression by which the maximum is determined is dependent on the architecture,
-         * and is something we don't really want to copy to userspace, as it is dependent on implementation
-         * details of the kernel. Since the kernel doesn't expose the maximum value to us, we can only try
-         * and hope. Hence, let's start with INT_MAX, and then keep halving the value until we find one that
-         * works. Ugly? Yes, absolutely, but kernel APIs are kernel APIs, so what do can we do... 🤯 */
+        /* cf. https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/file.c?h=v6.8#n27
+	 * for the progeny of the below value for `sysctl_nr_open_max`. Note
+	 * that the below logic was first introduced in `git` commit
+	 * eceea0b3df05ed262ae32e0c6340cc7a3626632d of the `linux` kernel and
+	 * was designed to prevent overflows when determining the maximum
+	 * number of possible file descriptors. */
+	#define BITS_PER_LONG __WORDSIZE
+        #define __const_min(x, y) ((x) < (y) ? (x) : (y))
+        unsigned sysctl_nr_open_max =
+            __const_min(INT_MAX, ~(size_t)0/sizeof(void *)) & -BITS_PER_LONG;
+
+        unsigned v = sysctl_nr_open_max;
 
         for (;;) {
-                int k;
-
-                v &= ~(__SIZEOF_POINTER__ - 1); /* Round down to next multiple of the pointer size */
                 if (v < 1024) {
                         log_warning("Can't bump fs.nr_open, value too small.");
                         break;
                 }
 
-                k = read_nr_open();
-                if (k < 0) {
-                        log_error_errno(k, "Failed to read fs.nr_open: %m");
+                int nr_open = read_nr_open();
+                if (nr_open < 0) {
+                        log_error_errno(nr_open, "Failed to read fs.nr_open: %m");
```

Comment:
Not necessary to update here.

Reply:
This would change the behavior relative to what exists today in `master`: https://github.com/systemd/systemd/blob/ce78bae2b85c526f189bee019b770b3e07f5aff1/src/core/main.c#L1274-L1278 . Are you saying you want me to pull this file read above the loop? I can do that, but then this PR does more than simply change the default value of `v` and, potentially (although unlikely), introduces a different failure mode than this proposed change set. I'd prefer that work to be separated into a different change set (PR).

Reply:
Okay, maybe I understand what you meant. I submitted f1f10b0d7a8c4c1e94a2f1debcd4a3ce4f693648 to try to address this.